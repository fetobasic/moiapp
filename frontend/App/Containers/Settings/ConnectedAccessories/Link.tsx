import React, { useState } from 'react';
import { StyleSheet, Text, View } from 'react-native';
import { ApplicationStyles, Colors, Fonts, Metrics } from 'App/Themes';
import { Button, HeaderSimple as Header, InformationRow as Row, Pressable } from 'App/Components';
import { NativeStackScreenProps } from '@react-navigation/native-stack';
import { SettingsStackParamList } from 'App/Types/NavigationStackParamList';
import { navigationGoBack } from 'App/Navigation/AppNavigation';
import { isTankMode } from 'App/Services/Accessory';
import { update } from 'App/Services/ConnectionControler';
import { devicesActions } from 'App/Store/Devices';
import { useDispatch } from 'react-redux';
import { ForeignAcsry, YetiState } from 'App/Types/Yeti';

type Props = NativeStackScreenProps<SettingsStackParamList, 'Link'>;

const Link = ({ route }: Props) => {
  const device = route.params.device as YetiState;
  const [mode, setMode] = useState<number>(device?.foreignAcsry?.mode || 0);
  const [isSettingChanged, setChanged] = useState<boolean>(false);
  const dispatch = useDispatch();

  const onChangeModeHandler = () => {
    setMode(isTankMode(mode) ? 1 : 0);
    setChanged(true);
  };

  const onSaveHandler = () => {
    // @ts-ignore TODO: we need to use complex and recursive Partial generic of stateObject type
    update({
      thingName: device.thingName || '',
      stateObject: { state: { desired: { foreignAcsry: { mode } as ForeignAcsry } } },
      isDirectConnection: Boolean(device?.isDirectConnection),
      updateDeviceState: (thingName, data) => dispatch(devicesActions.devicesAddUpdate.request({ thingName, data })),
    });
    navigationGoBack();
  };

  return (
    <View style={ApplicationStyles.mainContainer}>
      <Header title="Link" isChangeSaved={isSettingChanged} cbSave={onSaveHandler} />
      <View style={styles.container}>
        <Row
          style={styles.infoRow}
          textStyle={styles.textStyle}
          title="Link"
          addLeftIcon={<Text style={styles.linkIcon}>LINK</Text>}
          subTitle="The LINK connection is shown when a Link is detected on the expansion port."
          trimSubTitle={false}
        />
        <View style={styles.btnWrapper}>
          <Pressable
            style={[styles.btn, { backgroundColor: isTankMode(mode) ? Colors.green : Colors.background }]}
            onPress={onChangeModeHandler}>
            <Text style={[Fonts.font.base.button, { color: isTankMode(mode) ? Colors.black : Colors.green }]}>
              Tank Mode
            </Text>
          </Pressable>
          <Pressable
            style={[styles.btn, { backgroundColor: isTankMode(mode) ? Colors.background : Colors.green }]}
            onPress={onChangeModeHandler}>
            <Text style={[Fonts.font.base.button, { color: isTankMode(mode) ? Colors.green : Colors.black }]}>
              Car Mode
            </Text>
          </Pressable>
        </View>
        <Row
          title="Firmware Version"
          subTitle="Updated by Yeti."
          description={`Version ${device?.foreignAcsry?.firmwareVersion || 0}`}
        />
      </View>
      <Button
        style={styles.saveBtn}
        title="SAVE"
        onPress={onSaveHandler}
        disabled={!isSettingChanged}
        inverse={!isSettingChanged}
      />
    </View>
  );
};

export default Link;

const styles = StyleSheet.create({
  container: {
    paddingHorizontal: Metrics.baseMargin,
    paddingVertical: Metrics.marginSection,
    flex: 1,
  },
  infoRow: {
    marginTop: Metrics.halfMargin,
  },
  linkIcon: {
    ...Fonts.font.base.subtitleTwo,
    marginRight: 7,
    marginTop: 4,
  },
  textStyle: {
    marginLeft: 10,
  },
  btnContainer: {
    borderRadius: 24,
  },
  btnWrapper: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    backgroundColor: Colors.background,
    borderRadius: 24,
    shadowColor: Colors.black,
    shadowOpacity: 0.2,
    shadowOffset: { width: 4, height: 4 },
    shadowRadius: 12,
    elevation: 12,
    marginVertical: Metrics.halfMargin,
  },
  btn: {
    borderRadius: 24,
    height: 48,
    width: '50%',
    alignItems: 'center',
    justifyContent: 'center',
    color: Colors.green,
  },
  saveBtn: {
    margin: Metrics.baseMargin,
  },
});
