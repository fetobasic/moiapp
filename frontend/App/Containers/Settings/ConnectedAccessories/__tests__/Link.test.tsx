import React from 'react';
import { render } from '@testing-library/react-native';
import { NavigationContainer } from '@react-navigation/native';
import { Provider } from 'react-redux';
import createMockStore from 'redux-mock-store';
import { mockedState } from 'App/Fixtures/mocks/mockedState';
import Link from '../Link';
import { yeti6GState } from 'App/Fixtures/yeti6GState';

const mockStore = createMockStore([]);
const store = mockStore({
  ...mockedState,
});

describe('Link', () => {
  test('renders screen', () => {
    const props = {
      navigation: {
        navigate: jest.fn() as any,
      },
      route: { params: { device: yeti6GState } },
    } as any;

    const tree = render(
      <Provider store={store}>
        <NavigationContainer>
          <Link {...props} />
        </NavigationContainer>
      </Provider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
