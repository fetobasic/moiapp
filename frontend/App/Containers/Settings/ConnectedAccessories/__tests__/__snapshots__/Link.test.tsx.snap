// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link renders screen 1`] = `
<View
  style={
    {
      "backgroundColor": "rgba(27, 27, 27, 1)",
      "flex": 1,
    }
  }
>
  <View
    style={
      [
        {
          "backgroundColor": "rgba(27, 27, 27, 1)",
          "borderBottomLeftRadius": 24,
          "borderBottomRightRadius": 24,
          "elevation": 12,
          "shadowColor": "rgba(0, 0, 0, 1)",
          "shadowOffset": {
            "height": 4,
            "width": 4,
          },
          "shadowOpacity": 0.2,
          "shadowRadius": 12,
          "zIndex": 1,
        },
        {
          "paddingTop": 0,
        },
      ]
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "height": 60,
          "justifyContent": "center",
          "marginHorizontal": 24,
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        hitSlop={20}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "opacity": 1,
            },
            {
              "left": 0,
              "padding": 10,
              "position": "absolute",
            },
          ]
        }
      >
        <SvgMock />
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "marginHorizontal": 48,
          }
        }
      >
        <Text
          numberOfLines={1}
          style={
            {
              "color": "rgba(255, 255, 255, 1)",
              "fontFamily": "RobotoCondensed-Medium",
              "fontSize": 20,
              "fontWeight": "500",
              "lineHeight": 32,
            }
          }
        >
          Link
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      {
        "flex": 1,
        "paddingHorizontal": 24,
        "paddingVertical": 16,
      }
    }
  >
    <View
      style={
        [
          {
            "paddingHorizontal": 12,
            "paddingVertical": 15,
          },
          {
            "marginTop": 12,
          },
        ]
      }
    >
      <View
        style={
          [
            {
              "flexDirection": "row",
              "justifyContent": "space-between",
            },
            undefined,
          ]
        }
      >
        <View
          style={
            [
              {
                "marginLeft": 8,
              },
              undefined,
            ]
          }
        >
          <Text
            style={
              {
                "color": "rgba(255, 255, 255, 1)",
                "fontFamily": "Roboto-Medium",
                "fontSize": 14,
                "fontWeight": "500",
                "lineHeight": 20,
                "marginRight": 7,
                "marginTop": 4,
              }
            }
          >
            LINK
          </Text>
        </View>
        <View
          style={
            [
              {
                "flex": 1,
                "justifyContent": "center",
              },
              {
                "marginLeft": 10,
              },
            ]
          }
        >
          <Text
            ellipsizeMode="tail"
            numberOfLines={1}
            style={
              [
                {
                  "color": "rgba(255, 255, 255, 0.87)",
                  "fontFamily": "Roboto-Medium",
                  "fontSize": 16,
                  "fontWeight": "400",
                  "lineHeight": 24,
                  "maxWidth": 210,
                },
                undefined,
              ]
            }
          >
            Link
          </Text>
          <Text
            ellipsizeMode="tail"
            style={
              [
                {
                  "color": "rgba(147, 147, 147, 1)",
                  "fontFamily": "Roboto-Medium",
                  "fontSize": 14,
                  "fontWeight": "400",
                  "lineHeight": 20,
                },
                undefined,
                undefined,
              ]
            }
          >
            The LINK connection is shown when a Link is detected on the expansion port.
          </Text>
        </View>
        <View
          style={
            [
              {
                "alignItems": "center",
                "flexDirection": "row",
              },
              {
                "alignItems": "flex-start",
              },
            ]
          }
        />
      </View>
    </View>
    <View
      style={
        {
          "backgroundColor": "rgba(27, 27, 27, 1)",
          "borderRadius": 24,
          "elevation": 12,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginVertical": 12,
          "shadowColor": "rgba(0, 0, 0, 1)",
          "shadowOffset": {
            "height": 4,
            "width": 4,
          },
          "shadowOpacity": 0.2,
          "shadowRadius": 12,
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "opacity": 1,
            },
            [
              {
                "alignItems": "center",
                "borderRadius": 24,
                "color": "rgba(191, 210, 43, 1)",
                "height": 48,
                "justifyContent": "center",
                "width": "50%",
              },
              {
                "backgroundColor": "rgba(191, 210, 43, 1)",
              },
            ],
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "rgba(255, 255, 255, 1)",
                "fontFamily": "Roboto-Medium",
                "fontSize": 14,
                "fontWeight": "700",
                "lineHeight": 20,
                "textTransform": "uppercase",
              },
              {
                "color": "rgba(0, 0, 0, 1)",
              },
            ]
          }
        >
          Tank Mode
        </Text>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "opacity": 1,
            },
            [
              {
                "alignItems": "center",
                "borderRadius": 24,
                "color": "rgba(191, 210, 43, 1)",
                "height": 48,
                "justifyContent": "center",
                "width": "50%",
              },
              {
                "backgroundColor": "rgba(27, 27, 27, 1)",
              },
            ],
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "rgba(255, 255, 255, 1)",
                "fontFamily": "Roboto-Medium",
                "fontSize": 14,
                "fontWeight": "700",
                "lineHeight": 20,
                "textTransform": "uppercase",
              },
              {
                "color": "rgba(191, 210, 43, 1)",
              },
            ]
          }
        >
          Car Mode
        </Text>
      </View>
    </View>
    <View
      style={
        [
          {
            "paddingHorizontal": 12,
            "paddingVertical": 15,
          },
          undefined,
        ]
      }
    >
      <View
        style={
          [
            {
              "flexDirection": "row",
              "justifyContent": "space-between",
            },
            undefined,
          ]
        }
      >
        <View
          style={
            [
              {
                "flex": 1,
                "justifyContent": "center",
              },
              undefined,
            ]
          }
        >
          <Text
            ellipsizeMode="tail"
            numberOfLines={1}
            style={
              [
                {
                  "color": "rgba(255, 255, 255, 0.87)",
                  "fontFamily": "Roboto-Medium",
                  "fontSize": 16,
                  "fontWeight": "400",
                  "lineHeight": 24,
                  "maxWidth": 210,
                },
                undefined,
              ]
            }
          >
            Firmware Version
          </Text>
          <Text
            ellipsizeMode="tail"
            numberOfLines={1}
            style={
              [
                {
                  "color": "rgba(147, 147, 147, 1)",
                  "fontFamily": "Roboto-Medium",
                  "fontSize": 14,
                  "fontWeight": "400",
                  "lineHeight": 20,
                },
                undefined,
                undefined,
              ]
            }
          >
            Updated by Yeti.
          </Text>
        </View>
        <View
          style={
            [
              {
                "alignItems": "center",
                "flexDirection": "row",
              },
              {
                "alignItems": "flex-start",
              },
            ]
          }
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              [
                {
                  "opacity": 1,
                },
                undefined,
              ]
            }
          >
            <Text
              ellipsizeMode="clip"
              numberOfLines={1}
              style={
                [
                  {
                    "color": "rgba(255, 255, 255, 0.87)",
                    "fontFamily": "Roboto-Medium",
                    "fontSize": 16,
                    "fontWeight": "400",
                    "lineHeight": 24,
                    "maxWidth": 210,
                  },
                  undefined,
                ]
              }
            >
              Version 0
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": true,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "opacity": 1,
        },
        [
          {
            "elevation": 12,
            "shadowColor": "rgba(0, 0, 0, 1)",
            "shadowOffset": {
              "height": 4,
              "width": 4,
            },
            "shadowOpacity": 0.2,
            "shadowRadius": 12,
          },
          {
            "margin": 24,
          },
        ],
      ]
    }
  >
    <View
      style={
        {
          "borderRadius": 24,
          "overflow": "hidden",
        }
      }
    >
      <RNSVGSvgView
        bbHeight="48"
        bbWidth="100%"
        focusable={false}
        height={48}
        style={
          [
            {
              "backgroundColor": "transparent",
              "borderWidth": 0,
            },
            {
              "flex": 0,
              "height": 48,
              "width": "100%",
            },
          ]
        }
        width="100%"
      >
        <RNSVGGroup
          fill={
            {
              "payload": 4278190080,
              "type": 0,
            }
          }
        >
          <RNSVGDefs>
            <RNSVGLinearGradient
              gradient={
                [
                  0,
                  -13224394,
                  1,
                  1776411,
                ]
              }
              gradientTransform={null}
              gradientUnits={0}
              name="grad"
              x1="0"
              x2="0"
              y1="0"
              y2="0.5"
            />
          </RNSVGDefs>
          <RNSVGRect
            fill={
              {
                "brushRef": "grad",
                "type": 1,
              }
            }
            height="48"
            propList={
              [
                "fill",
              ]
            }
            width="100%"
            x="0"
            y="0"
          />
          <View
            style={
              [
                {
                  "borderRadius": 24,
                  "marginHorizontal": 1,
                  "marginTop": 1,
                },
                {
                  "backgroundColor": "rgba(27, 27, 27, 1)",
                  "height": 46,
                },
              ]
            }
          >
            <View
              style={
                [
                  {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "center",
                    "paddingVertical": 12,
                  },
                  undefined,
                ]
              }
            >
              <View
                style={
                  {
                    "alignItems": "center",
                    "flexDirection": "row",
                  }
                }
              >
                <Text
                  style={
                    [
                      {
                        "color": "rgba(255, 255, 255, 1)",
                        "fontFamily": "Roboto-Medium",
                        "fontSize": 14,
                        "fontWeight": "700",
                        "lineHeight": 20,
                        "textTransform": "uppercase",
                      },
                      {
                        "color": "rgba(255, 255, 255, 0.38)",
                      },
                      undefined,
                    ]
                  }
                >
                  SAVE
                </Text>
              </View>
            </View>
          </View>
        </RNSVGGroup>
      </RNSVGSvgView>
    </View>
  </View>
</View>
`;
